### PR TITLE
Refactor locale&string&template related code

### PR DESCRIPTION
**Pull request by wxiaoguang**
Clarify when "string" should be used (and be escaped), and when "template.HTML" should be used (no need to escape)

And help PRs like  #29059 , to render the error messages correctly.

The only possible regression bug is that something on the UI might get double-escaped, which is pretty easy to fix (and I will fix them in first time, as always). Update: if such regression happens, it should be the old code's bug, which needs to be fixed accordingly.

### CODE REVIEW

File: modules/context/context.go [hunk 1]
```diff
@@ -253,6 +243,12 @@ func (ctx *Context) JSONOK() {
...
 }
 
-func (ctx *Context) JSONError(msg string) {
-	ctx.JSON(http.StatusBadRequest, map[string]any{"errorMessage": msg})
+func (ctx *Context) JSONError(msg any) {
+	switch v := msg.(type) {
+	case string:
+		ctx.JSON(http.StatusBadRequest, map[string]any{"errorMessage": v, "renderFormat": "text"})
+	case template.HTML:
+		ctx.JSON(http.StatusBadRequest, map[string]any{"errorMessage": v, "renderFormat": "html"})
```
Comment (lunny) [inline modules/context/context.go:251]: return?

Comment (wxiaoguang) [reply modules/context/context.go:251]: ~~why "return" is needed?~~ 

Oops .... I see

Comment (wxiaoguang) [reply modules/context/context.go:251]: Done in 3a5c9a757126bf309f435b9f15e335b5507b7c02

**Comment by KN4CK3R** [comment]
Wouldn't it be easier to add a `TrHtml` method? Then you don't need most changes and we have no potential problem with current PRs which would use the wrong method.

**Comment by wxiaoguang** [comment]
> Wouldn't it be easier to add a `TrHtml` method? Then you don't need most changes and we have no potential problem with current PRs which would use the wrong method.

No, the real problem is that `Tr` should return `template.HTML`.

File: routers/web/repo/view.go [hunk 2]
```diff
@@ -758,7 +759,7 @@ func checkHomeCodeViewable(ctx *context.Context) {
 		}
 	}
 
-	ctx.NotFound("Home", fmt.Errorf(ctx.Tr("units.error.no_unit_allowed_repo")))
+	ctx.NotFound("Home", errors.New("units.error.no_unit_allowed_repo"))
```
Comment (lunny) [inline routers/web/repo/view.go]: No traslation now?

Comment (wxiaoguang) [reply routers/web/repo/view.go]: The second argument is for "log error". It doesn't make sense to output translated message into log.

It is the old code's problem.

Comment (wxiaoguang) [reply routers/web/repo/view.go]: Reverted to old code by 32f0cd31729c4bb4da242c3fc20c6a1258031029

File: modules/auth/password/password.go [hunk 3]
```diff
@@ -121,15 +122,15 @@ func Generate(n int) (string, error) {
 }
 
 // BuildComplexityError builds the error message when password complexity checks fail
-func BuildComplexityError(locale translation.Locale) string {
+func BuildComplexityError(locale translation.Locale) template.HTML {
```
Comment (lunny) [inline modules/auth/password/password.go:125]: `flashMsgStringOrHTML` will convert the output to a string. Is that expected?

Comment (wxiaoguang) [reply modules/auth/password/password.go:125]: Yes, that's the old behavior of the "Flash" package. The "message" will be rendered by `Str2html`, so there is no problem at the moment. See the `alert.tmpl`.

----

And more details: there is only type-casting (HTML -> string -> HTML), no escaping/unescaping, so no problem.

**Review by lunny — APPROVED** [review]

**Comment by delvh** [comment]
I don't understand how this makes things safer, for me it is less safe…

**Comment by delvh** [comment]
Yep, I've noticed what you mean by looking further through your PR.
Now, we simply penalize having safe parameters by potentially double escaping them.
I don't like this write overhead for safe parameters, but it is much better than a security vulnerability.

**Comment by wxiaoguang** [comment]
> I don't like this write overhead for safe parameters, but it is much better than a security vulnerability.

There is no difference (comparing to before). Suppose there is a "safe" variable `$var1` and a normal variable `$var2`

Before: `{{ctx.Locale.Tr $var1 $var2 | Safe}}` , well , it still needs a "|Safe" but the "|Escape" could be easily forgotten on `$var2`. The correct code should be `{{ctx.Locale.Tr $var1 ($var2|Escape) | Safe}}` 

After: `{{ctx.Locale.Tr ($var1|Safe) $var2}}` , everything works as expected. Isn't it much better?

File: templates/repo/issue/view_title.tmpl [hunk 4]
```diff
@@ -56,18 +56,18 @@
 					{{$mergedStr:= TimeSinceUnix .Issue.PullRequest.MergedUnix ctx.Locale}}
 					{{if .Issue.OriginalAuthor}}
 						{{.Issue.OriginalAuthor}}
-						<span class="pull-desc">{{ctx.Locale.Tr "repo.pulls.merged_title_desc" .NumCommits $headHref $baseHref $mergedStr | Safe}}</span>
+						<span class="pull-desc">{{ctx.Locale.Tr "repo.pulls.merged_title_desc" .NumCommits ($headHref|Safe) ($baseHref|Safe) $mergedStr}}</span>
```
Comment (delvh) [inline templates/repo/issue/view_title.tmpl:59]: How are timestamps handled?
Are they escaped twice now, or did they already return `template.HTML`?

File: routers/web/auth/password.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ func ForgotPasswd(ctx *context.Context) {
 	ctx.Data["Title"] = ctx.Tr("auth.forgot_password_title")
 
 	if setting.MailService == nil {
-		log.Warn(ctx.Tr("auth.disable_forgot_password_mail_admin"))
+		log.Warn("no mail service configured")
```
Comment (delvh) [inline routers/web/auth/password.go:40]: :+1:

**Review by delvh — APPROVED** [review]

File: templates/repo/issue/view_title.tmpl (hunk 4 above)
Comment (wxiaoguang) [reply templates/repo/issue/view_title.tmpl:59]: > or did they already return `template.HTML`?

Yes, they are already `template.HTML`, see function `TimeSinceUnix`

**Comment by wxiaoguang** [comment]
Let's try this, and if there is anything wrong, I will fix them asap.

**Comment by delvh** [comment]
Backport?

**Comment by wxiaoguang** [comment]
> Backport?

No idea (maybe I slightly prefer no), it doesn't seems easy to backport. If most people would like to, I could try.

**Pull request merged**